Add interpretation of slow test tags

emord opened the pull request:
@millerdev @snopoke @calellowitz @dannyroberts @nickpell I think you all participated in the testing side chat. This is my suggestion going forward.

We currently have a plugin that tracks test times. Past a [certain threshold](https://github.com/dimagi/commcare-hq/blob/master/.travis.yml#L42) we log the timings to the travis log file. This PR would introduce [tagging a test](https://docs.djangoproject.com/en/1.11/topics/testing/tools/#tagging-tests) with 'slow_X` where X is seconds. If a test is tagged with  that, it is allowed to run that number of seconds. Otherwise if it runs over the threshold, it fails. Going forward I'd like to:

1. Go back through travis test run logs and tag all tests in this PR that are > 10 seconds with `slow_X`
2. Merge this PR
3. Periodically reduce the threshold and tag all tests that begin to fail.
4. Keep going until we hit 1s or less

This hopefully could prevent new, large tests from entering the codebase and more explicitly document what tests are ripe to be refactored and examined more closely.

Some issues I can see coming up:

- I think there are instances where the setup time bleeds into test times (particularly migrations). This might add some noisy tagging on otherwise ok tests.
- Would be nice to also track time it takes to migrate databases
- This might discourage people form making tests.
- This only solves adding new large tests, not adding 1,000 kind of slow tests. I'm hoping that if its death by 1,000 cuts at least we'll get 1,000 tests out of it

emord committed:
Add interpretation of slow test tags

dannyroberts commented:
This a pretty cool idea. My only comment is about the travis log and/or failure output, that we could have those support the process of tagging with some thoughtful formatting.

The failure message might say

```
Test ran in 15 seconds and is greater than threshold 2. To fix either speed up the test or add
@tag('slow_17') to <path.to.test:TestClass.test_func>
```

(17 is 10% more than 15, rounded up. Getting the right buffer here could prevent this from being super brittle.)

Additionally since you were suggesting a plan of repeatedly dropping the threshold, I think it would make sense to keep a logging threshold that's lower than the failure threshold. The logging threshold would be set to the next target failure threshold, and it would have similiarly nice output, printing a suggested `@tag('slow_X')` decorator. That would make that person's job easier—they can just pick a recent build and grep it for `tag..slow_` (or whatever) and have a list of tests to look at, and what tags to add to the ones they can't improve.

calellowitz commented:
Generally very on board with this idea. Is this only for individual tests, i know you mentioned setup as well on slack, but I am curious if that would create some perverse incentives around where work should happen. In theory doing everything in the module/class setup is preferred because it only happens once, but if that is capped in time and tests are otherwise fast, it may be easier to push pieces to repeated test setup than bring down module setup by other means, which creates on balance longer tests, but no component above our thresholds. Obviously, if only tests are being timed, this does not apply.

emord commented:
> or add @tag('slow_17') to <path.to.test:TestClass.test_func>

I think this could be useful for the interim to when going back through and tagging slow tests. However I think we should have sufficiently high barrier to entry of adding slow tests that I'd be hesitant to have an automated message telling you that its an option.

> Is this only for individual tests

I think that the current timing plugin splits up running, setup and teardown (but i'm not sure offhand how setupclass vs setup would work)

> that would create some perverse incentives around where work should happen

Agreed. I think this is a much better way of saying what I was thinking when I mentioned discouraging people to write tests. which is to say that it could discourage people from writing performant tests in the correct way if the plugin is not capturing/catching the bad behavior correctly

millerdev commented on custom/aaa/tests/test_utils.py:
```diff
@@ -333,4 +337,7 @@ def test_build_location_filters_for_state_without_child_MWCD(self):
+        self.assertDictEqual(result, expected_output)
+
+    def slow_test(self):
+        time.sleep(5)
```
Is this intended to be part of the final state of this PR? Or maybe you were just verifying that the timing plugin works the way you expected and were planning to remove it?

emord replied:
Correct. This wouldnt be in the final PR. just using it for an example testing

millerdev reviewed (commented):
I'm very much in favor of the idea of making slow tests more visible, but also agree with others that we should be careful about setting up a system that encourages bad behaviors. Can we set this up in a way that we can monitor for abuse and hopefully curb it before it gets out of hand?

It's a little disappointing that by adding the time to the tag value (`slow_N`) we make it much harder if not impossible to run tests that are not "slow" (`-a '!slow'`). What do you think of using nose's way more powerful [`@attr` decorator](https://nose.readthedocs.io/en/latest/plugins/attrib.html) rather than Django's `@tag`? Then we could do something like `@attr(slow=10)`? We'd even get fancy things like `--eval-attr='slow < 5'` for free.

emord commented:
> Can we set this up in a way that we can monitor for abuse and hopefully curb it before it gets out of hand?

My initial thoughts were code review could be the place where this is monitored. As in don't accept PRs with weird tests. Although I hope we can trust each other to not try to game the system intentionally. This is mostly trying to put the alerting that you're adding a slow test before tests get too bad.

I like nose's attr. didn't know about it before. Will look into ittomorrow

snopoke commented:
Nothing to add that hasn't already been mentioned. Overall I like the idea.

emord committed:
Change django tags to nose attrs

emord committed:
Document slow tests from first job

emord committed:
Document slow tests from second job

emord committed:
Document slow tests from third job

emord committed:
Document slow tests from fourth job

emord committed:
Remove errors for 'before' event

emord committed:
Add link to testing documentation

emord commented:
went through a recent test run and looked for strings "setup,", "teardown," and "run," to find slow tests, and added some buffer to them

millerdev commented on corehq/apps/hqwebapp/tests/test_requirejs.py:
```diff
@@ -48,6 +50,7 @@ def _categorize_file(filename):
 
         cls._run_jobs(cls.js_files, _categorize_file)
 
+    @attr(slow=16)  # due to setup
```
Would be nice if we could tag the setup method rather than a random test on the class with slow setup.

emord replied:
Agreed. I'm not sure how easy that will be. Another thing I've run into is that since we skip a lot of tests in our divided tests, it's not clear where that time is getting allocated. May need to think about this more if it becomes an issue, for now I'm ok with commenting about it

millerdev commented on corehq/apps/accounting/tests/test_autopay.py:
```diff
@@ -19,6 +20,7 @@
 
 
+@attr(slow=700)  # almost certainly migrations
 class TestBillingAutoPay(BaseInvoiceTestCase):
```
Is this test associated with migrations because it is the first test to run after migrations are applied? That seems pretty fragile—I think adding a new test with a module path that sorts alphabetically before this one would break it. Would be great if we could time the migrations separately so their time does not get added to some random test.

emord replied:
Agreed that this is fragile. I didn't want to complicate the changes to the plugin too much. Another issue with this is that I have no idea how fast these are.

Another way to work around this could be to proactively create that test case with a slow attr?

millerdev commented on corehq/tests/noseplugins/timing.py:
```diff
@@ -114,3 +121,16 @@ def end_event(name, context):
+def _get_expected_slow_seconds(test_context):
+    # class is tagged for class based test
+    # For function based test, the function is tagged
+    slow_seconds = getattr(test_context, 'slow', None)
```
If the class is tagged, does that time apply to each test? For example, if a class having three test methods is tagged `slow=15` then each test gets a timeout of 15s for a total of 45s overall? Or is it 15s total for the whole class?

emord replied:
It applies to each class

millerdev commented on corehq/tests/noseplugins/timing.py:
```diff
@@ -114,3 +121,16 @@ def end_event(name, context):
     PLUGIN_INSTANCE.end_event(name, context)
+
+
+def _get_expected_slow_seconds(test_context):
```
Does this work for stand-alone `def test_xyz():` function tests that are not part of a class? ([example](https://github.com/dimagi/commcare-hq/blob/dm/pillow-load/corehq/util/tests/test_celery_utils.py#L12))

emord replied:
That did not work. updated https://github.com/dimagi/commcare-hq/pull/23874/commits/9d4d82d933f99490fb328f3961e532d346510739

millerdev reviewed (commented):
This is looking great!

emord committed:
Remove example slow test

emord committed:
Fix slow function tests

emord committed:
Catch tests that we skip, but are attached to general test setup

emord committed:
Add more slow tags

emord committed:
Remove --stop to get all slow tests

emord committed:
Merge branch 'master' into je/test-timings

millerdev commented on .travis.yml:
```diff
@@ -39,7 +39,7 @@ script:
       bash scripts/test-make-requirements.sh
     fi
-  - "scripts/docker test --noinput --stop --verbosity=2 --divide-depth=1 --with-timing --threshold=10"
+  - "scripts/docker test --noinput --verbosity=2 --divide-depth=1 --with-timing --threshold=10"
```
Will have a negative impact on queue throughput. Is this just a temporary change to find slow tests (to be reverted later before this PR is merged), or are you intending to keep it indefinitely?

emord replied:
yep this is just temporary and it doesn't work that well anywyas to some test cleanup issues i think.

I'm probably going to lower the threshold for logging to get more logs at once and give these tests some more buffer. They appear to be more variable that i thought

emord committed:
Do not fail slow tests by default

emord committed:
Add --stop to travis

emord committed:
Lower threshold to 1 second

millerdev reviewed (commented):
Was the langcodes submodule update in 0b93ef7205a32887617585694e62ce486bc6348b intentional?

emord committed:
Merge branch 'master' into je/test-timings

emord committed:
Couple of more slow tests

emord committed:
Change paramater based on test event

emord committed:
Tag more tests

emord committed:
Fail tests past 10 seconds

emord committed:
Update .travis.yml

emord committed:
Add docs on slow test attr

emord committed:
Merge branch 'master' into je/test-timings

emord commented:
@millerdev I think you've reviewed this the most, and I think this is ready for a final review. Changes I've made:

- Setup and teardown is tagged separately
- Changed the threshold to 11
- Went through previous test logs to view all tests that were > 8 seconds and tagged them (Have found more variance on these than I expected
- Added docs to the testing page on rst (and linked it in the test failures)
- Added option to fail fast tests (by default it does not fail)

I think this should be enough to make these tests not fragile, but if we do find that existing tests begin failing at a high rate, we can remove the option to fail temporarily

millerdev commented on corehq/tests/noseplugins/timing.py:
```diff
@@ -114,3 +125,32 @@ def end_event(name, context):
+
+def _get_expected_slow_seconds(event, test_context):
+    attr_name = {
+        "before": "slow_before",
```
Doesn't look like this is used anywhere (or [even documented](https://github.com/dimagi/commcare-hq/pull/23874/commits/06577972f57141e71af205e78daa62562723f334)). Do you think it would make sense to remove it until we need it? If not, what do you think of documenting it, clearly distinguishing the difference between `slow_setup` and `slow_before`?

emord replied:
https://github.com/dimagi/commcare-hq/pull/23874/commits/6bbb8e6172c77ffa0220b05e31fb545bfe3b6451

emord committed:
Merge branch 'master' into je/test-timings

emord committed:
Remove undocumented/unused tag

orangejenny commented on corehq/apps/hqwebapp/tests/test_requirejs.py:
```diff
@@ -94,6 +97,7 @@ def _test_file(filename):
 
 
 class TestRequireJSBuild(SimpleTestCase):
+    @attr(slow=40)
```
this test doesn't exist anymore!
